src/lib.rs

```diff
@@ -218,10 +212,166 @@ use srml_support::traits::Currency;
+struct ModerationAction<BlockNumber, Moment, AccountId> {
+
+    /// When action occured.
+    occured_at: BlockchainTimestamp<BlockNumber, Moment>,
+
+    /// Account forum sudo which acted.
+    forum_sudo_account_id: AccountId,
+
+    /// Moderation rationale
+    rationale: Vec<u8>
+
 }
 
+/// Represents a revision of the text of a Post
+struct PostTextEdit<BlockNumber, Moment> {
+
+    /// What this edit occured.
+    occured_at: BlockchainTimestamp<BlockNumber, Moment>,
+
+    /// New text
+    new_text: Vec<u8>
+}
+
+/// Represents a thread post
+type PostId = u64;
+struct Post<BlockNumber, Moment, AccountId> {
+
+    /// Post identifier
+    id: PostId,
+
+    /// Id of thread to which this post corresponds.
+    thread_id: ThreadId,
+
+    // Position of post in thread.
+    position: u64,
+
+    /// Initial text of post
+    initial_text: Vec<u8>,
+
+    /// Possible moderation of this post
+    moderation : Option<ModerationAction<BlockNumber, Moment, AccountId>>,
+    
+    /// Edits of post ordered chronologically by edit time.
+    post_text_edits: Vec<PostTextEdit<BlockNumber, Moment>>,
+
+    /// When post was submitted.
+    created_at : BlockchainTimestamp<BlockNumber, Moment>,
+    
+    /// Author of post.
+    author_id : AccountId
+
+}
+
+/// Represents a thread
+type ThreadId = u64;
+struct Thread<BlockNumber, Moment, AccountId> {
+
+    /// Thread identifier
+    id : ThreadId,
+
+    /// Title
+    title : Vec<u8>,
+
+    /// Category in which this thread lives
+    category: CategoryId,
+
+    /// Position of thread in category.
+    position: u64,
+
+    /// Possible moderation of this thread
+    moderation : Option<ModerationAction<BlockNumber, Moment, AccountId>>,
+
+    /// Number of unmoderated posts in this thread
+    num_unmoderated_posts: u64,
+
+    /// Number of moderated posts in this thread
+    num_moderated_posts: u64,
+
+    /// When thread was established.
+    created_at : BlockchainTimestamp<BlockNumber, Moment>,
+    
+    /// Author of post.
+    author_id : AccountId
+}
+
+/// Represents a category
+type CategoryId = u64;
+struct Category<BlockNumber, Moment, AccountId> {
+
+    /// Category identifier
+    id : CategoryId,
+
+    /// Title
+    title : Vec<u8>,
+
+    /// Description
+    description: Vec<u8>,
+
+    /// When category was established.
+    created_at : BlockchainTimestamp<BlockNumber, Moment>,
+
+    /// Whether category is deleted.
+    deleted: bool,
+
+    /// Whether category is archved.
+    archived: bool,
+
+    /// Number direct subcategories.
+    num_direct_subcategories: u64,
+
+    /// Number of unmoderated threads directly in this category.
+    num_direct_unmoderated_threads: u64,
+
+    /// Number of moderated threads directly in this category.
+    num_direct_moderated_threads: u64,
+
+    /// Parent category, if present, otherwise this category is a root categoryl
+    parent_category: Option<CategoryId>,
```

Comment:
Inconsistency in naming:
```suggestion
    parent_category_id: Option<CategoryId>,
```

Reply:
Fixed.